crypto/aes/asm/aes-cfb-avx512.pl

```diff
@@ -0,0 +1,922 @@
+
+    mov \$1,$mask                     # build a mask with the least significant $left bits set
+    shlq %cl,$mask                    # $left is left shift counter
+    dec $mask                         # $mask is 2^$left-1
+    kmovq $mask,%k1
+
+    mov $num,%rax                     # keep in-out num in %al
+    add $left,%rax                    # advance $num
+    and \$0x0F,%al                    # wrap-around $num in a 16-byte block
+
+    leaq ($num,$ivp),%r11             # process $left iv bytes
+    vmovdqu8 (%r11),%xmm0
+    vmovdqu8 ($inp),%xmm1             # process $left input bytes
+    vpxor %xmm0,%xmm1,%xmm2           # CipherFeedBack XOR
+    vmovdqu8 %xmm2,($out){%k1}        # write $left output bytes
+    vmovdqu8 %xmm1,(%r11){%k1}        # blend $left input bytes into iv
+
+    add $left,$inp                    # advance pointers
+    add $left,$out
+    sub $left,$len
+    jz .Laes_cfb128_dec_zero_pre      # return early if no AES encryption required
+
+    mov $key_backup,$key_original     # restore "key_original" as arg3
+
+.Laes_cfb128_dec_mid:
+___
+
+    &load_aes_key_schedule_4x();
+
+$code.=<<___;
+###########################################################
+# inner full blocks processing
+###########################################################
+
+    # $temp_4x is "iv | iv | iv | iv"
+    vbroadcasti32x4 ($ivp),$temp_4x             # load iv
+
+    cmp \$0x100,$len                            # are there 16 ciphertext blocks left (2048 bits) ?
+    jb .Laes_cfb128_dec_check_4x
+
+###########################################################
+# decrypt groups of 16 128-bit blocks in parallel
+# behaves as 16x loop unroll
+###########################################################
+
+.balign 32
+.Loop_aes_cfb128_dec_mid_16x:
+    sub \$0x100,$len
+
+    # load 16 ciphertext blocks
+
+    # $cipher_4x  is "ciphertext  0 | ciphertext  1 | ciphertext  2 | ciphertext  3"
+    vmovdqu32 ($inp),$cipher_4x
+    # $cipher_8x  is "ciphertext  4 | ciphertext  5 | ciphertext  6 | ciphertext  7"
+    vmovdqu32 64($inp),$cipher_8x
+    # $cipher_12x is "ciphertext  8 | ciphertext  9 | ciphertext 10 | ciphertext 11"
+    vmovdqu32 128($inp),$cipher_12x
+    # $cipher_16x is "ciphertext 12 | ciphertext 13 | ciphertext 14 | ciphertext 15"
+    vmovdqu32 192($inp),$cipher_16x
+
+    # $temp_4x  is   "iv            | ciphertext  0 | ciphertext  1 | ciphertext  2"
+    valignq \$6,$temp_4x,$cipher_4x,$temp_4x
+    # $temp_8x  is   "ciphertext  3 | ciphertext  4 | ciphertext  5 | ciphertext  6"
+    valignq \$6,$cipher_4x,$cipher_8x,$temp_8x
+    # $temp_12x is   "ciphertext  7 | ciphertext  8 | ciphertext  9 | ciphertext 10"
+    valignq \$6,$cipher_8x,$cipher_12x,$temp_12x
+    # $temp_16x is   "ciphertext 11 | ciphertext 12 | ciphertext 13 | ciphertext 14"
+    valignq \$6,$cipher_12x,$cipher_16x,$temp_16x
+
+    lea 256($inp),$inp                          # $inp points to next ciphertext
```

Comment:
Is there a reason for using 'lea' vs just an 'add' ?

Reply:
No particular reason, I expect no difference when replaced with `add`.